ml-agents/mlagents/trainers/ppo/multi_gpu_policy.py

```diff
@@ -84,6 +68,82 @@ def create_model(self, brain, trainer_params, reward_signal_configs, seed):
                         self.towers[-1].create_ppo_optimizer()
             self.model = self.towers[0]
 
+            avg_grads = self.average_gradients([t.grads for t in self.towers])
+            update_batch = self.model.optimizer.apply_gradients(avg_grads)
+
+            avg_value_loss = tf.reduce_mean(
+                tf.stack([model.value_loss for model in self.towers]), 0
+            )
+            avg_policy_loss = tf.reduce_mean(
+                tf.stack([model.policy_loss for model in self.towers]), 0
+            )
+
+        self.inference_dict.update(
+            {
+                "action": self.model.output,
+                "log_probs": self.model.all_log_probs,
+                "value": self.model.value_heads,
+                "entropy": self.model.entropy,
+                "learning_rate": self.model.learning_rate,
+            }
+        )
+        if self.use_continuous_act:
+            self.inference_dict["pre_action"] = self.model.output_pre
+        if self.use_recurrent:
+            self.inference_dict["memory_out"] = self.model.memory_out
+        if (
+            is_training
+            and self.use_vec_obs
+            and trainer_params["normalize"]
+            and not load
+        ):
+            self.inference_dict["update_mean"] = self.model.update_normalization
+
+        self.total_policy_loss = self.model.abs_policy_loss
+        self.update_dict.update(
+            {
+                "value_loss": avg_value_loss,
+                "policy_loss": avg_policy_loss,
+                "update_batch": update_batch,
+            }
+        )
+
+    def create_reward_signals(self, reward_signal_configs):
+        """
+        Create reward signals
+        :param reward_signal_configs: Reward signal config.
+        """
+        self.reward_signal_towers = []
+        with self.graph.as_default():
+            with tf.variable_scope(TOWER_SCOPE_NAME, reuse=tf.AUTO_REUSE):
+                for device_id, device in enumerate(self.devices):
+                    with tf.device(device):
+                        reward_tower = {}
+                        for reward_signal, config in reward_signal_configs.items():
+                            reward_tower[reward_signal] = create_reward_signal(
+                                self, self.towers[device_id], reward_signal, config
+                            )
+                            for k, v in reward_tower[reward_signal].update_dict.items():
+                                self.update_dict[k + "_" + str(device_id)] = v
+                        self.reward_signal_towers.append(reward_tower)
+                for _, reward_tower in self.reward_signal_towers[0].items():
+                    for _, update_key in reward_tower.stats_name_to_update_name.items():
+                        self.update_dict.update(
```

Comment:
Minor but this statement is 13 lines long.  I think we should have some intermediate variable names to make it easier to parse.